In Swift 3/4 mode, continue treating 'lazy override' as an override

jrose-apple opened the pull request:
Follow-up for #13304. Without this, the declaration would be accepted, but any uses of the overridden property would be treated as ambiguous because the property wouldn't really be marked as an override.

"Oops."

rdar://problem/35900345

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
And, to try to rule out me breaking anything subtle…

@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -6623,18 +6623,26 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+          // HACK: We would normally do this at the end of validation, but
+          // if we wait that long then the getter doesn't get marked as an
+          // override.
+          maybeAddAccessorsToVariable(cast<VarDecl>(overrideASD), TC);
```
Trivial getters are created in CodeSynthesis.cpp, maybe we should mark them as overrides there if the AbstractStorageDecl overrides something? Up until now this could not happen because trivial accessors are only created for stored properties and lazy properties which normally don't override anything.

jrose-apple replied:
Oh, that might make more sense. I could see `maybeAddAccessorsToVariable` depending on something earlier in validation.

slavapestov reviewed (commented):
This looks good, but you might want to follow up on the hack at some point.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Well, didn't get that right either. I'm not going to have time to work on this until the afternoon.

jrose-apple committed:
In Swift 3/4 mode, continue treating 'lazy override' as an override

Follow-up for 7c707ce97c9. Without this, the declaration would be
accepted, but any uses of the overridden property would be treated as
ambiguous because the property wouldn't really be marked as an
override.

rdar://problem/35900345

jrose-apple commented:
This is definitely becoming a bigger change than just `lazy`. Re-review, Slava?

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please smoke test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

slavapestov reviewed (approved):
LGTM!
